File: JavaScript/ex/CircularLinkedList.js
class nodeType {
  constructor(item) {
    this.data = item;
    this.link = null;
  }

  // 연결 리스트와 다른 점은 노드를 생성할 때 마지막 노드를 처음 노드에 연결 시켜주는 것이다.
  firstNodeIn(item) {
    let newNode = new nodeType(item);
    if (this.link === null) {
      this.link = newNode;
      newNode.link = this;
    } else {
      newNode.link = this.link;
      this.link = newNode;
    }
  }

  firstNodeOut() {
    if (this.link === null) return null;
    const tmp = this.link;
    this.link = tmp.link;
  }

  lastNodeIn(item) {
    let newNode = new nodeType(item);
    let p = this;
    if (this.link === null) {
      this.link = newNode;
      newNode.link = this;
    } else {
      while (p.link !== this) {
        p = p.link;
      }
      p.link = newNode;
      newNode.link = this;
    }
  }

  lastNodeOut() {
    if (this.link === null) return null;
    let p = this;
    while (p.link.link !== this) {
      p = p.link;
    }
    p.link = this;
  }

  print() {
    let string = `${this.data} | `;
    // for(let p = this.link; p != this; p=p.link){
    // }
    let p = this.link;
    while (p !== this) {
      string += `${p.data} | `;
      p = p.link;
    }
    string += p.data;
    console.log(string);
  }
}

File: Python/CircularLinkedList.py
class NodeType:
  def __init__(self, item):
    self.data = item
    self.link = None

  # 연결 리스트와 다른 점은 노드를 생성할 때 마지막 노드를 처음 노드에 연결 시켜주는 것이다.
  def first_node_in(self, item):
    new_node = NodeType(item)
    if self.link is None:
      self.link = new_node
      new_node.link = self
    else:
      new_node.link = self.link
      self.link = new_node

  def first_node_out(self):
    if self.link is None:
      return None
    tmp = self.link
    self.link = tmp.link

  def last_node_in(self, item):
    new_node = NodeType(item)
    p = self
    if self.link is None:
      self.link = new_node
      new_node.link = self
    else:
      while p.link is not self:
        p = p.link
      p.link = new_node
      new_node.link = self

  def last_node_out(self):
    if self.link is None:
      return None
    p = self
    while p.link.link is not self:
      p = p.link
    p.link = self

  def print(self):
    string = f"{self.data} | "
    p = self.link
    while p is not self:
      string += f"{p.data} | "
      p = p.link
    string += p.data
    print(string)
